src/ripple/app/paths/impl/FlowDebugInfo.h

```diff
@@ -0,0 +1,364 @@
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2012, 2013 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#ifndef RIPPLE_PATH_IMPL_FLOWDEBUGINFO_H_INCLUDED
+#define RIPPLE_PATH_IMPL_FLOWDEBUGINFO_H_INCLUDED
+
+#include <ripple/app/paths/impl/AmountSpec.h>
+#include <ripple/protocol/IOUAmount.h>
+#include <ripple/protocol/XRPAmount.h>
+
+#include <boost/container/flat_map.hpp>
+#include <boost/optional.hpp>
+
+#include <chrono>
+#include <sstream>
+
+namespace ripple
+{
+namespace path
+{
+namespace detail
+{
+// Track performance information of a single payment
+struct FlowDebugInfo
+{
+    using clock = std::chrono::high_resolution_clock;
+    using time_point = clock::time_point;
+    boost::container::flat_map<std::string, std::pair<time_point, time_point>>
+        timePoints;
+    boost::container::flat_map<std::string, std::size_t> counts;
+
+    struct PassInfo
+    {
+        PassInfo () = delete;
+        PassInfo (bool nativeIn_, bool nativeOut_)
+            : nativeIn (nativeIn_), nativeOut (nativeOut_)
+        {
+        }
+        bool nativeIn;
+        bool nativeOut;
+        std::vector<EitherAmount> in;
+        std::vector<EitherAmount> out;
+        std::vector<size_t> numActive;
+
+        std::vector<std::vector<EitherAmount>> liquiditySrcIn;
+        std::vector<std::vector<EitherAmount>> liquiditySrcOut;
+
+        void
+        reserve (size_t s)
+        {
+            in.reserve (s);
+            out.reserve (s);
+            liquiditySrcIn.reserve(s);
+            liquiditySrcOut.reserve(s);
+            numActive.reserve (s);
+        }
+
+        size_t
+        size () const
+        {
+            return in.size ();
+        }
+
+        void
+        push_back (EitherAmount const& in_amt,
+            EitherAmount const& out_amt,
+            std::size_t active)
+        {
+            in.push_back (in_amt);
+            out.push_back (out_amt);
+            numActive.push_back (active);
+        }
+
+        void
+        pushLiquiditySrc (EitherAmount const& in, EitherAmount const& out)
+        {
+            assert(!liquiditySrcIn.empty());
```

Comment:
@scottschurr Because it's a vector of vectors, one vector per pass. A new vector is pushed in `newLiquidityPass`. It is a logic error if `pushLiquiditySrc` is called before `newLiquidityPass`.